### ISSUE TITLE
Add a new environment variable which has the same value as `TRAVIS_BRANCH` or `TRAVIS_PULL_REQUEST_BRANCH` depending on the context

### ISSUE THREAD

**Issue by breerly** [travis-ci/travis-ci#6652]
The addition of `$TRAVIS_PULL_REQUEST_BRANCH` from https://github.com/travis-ci/travis-ci/issues/1633 is nice, but it is still a bit difficult to get the name of the _current_ branch name from both PUSH and PR builds.

The current workaround I have is this, which is not great: 

```
export BRANCH=$(if [ "$TRAVIS_PULL_REQUEST" == "false" ]; then echo $TRAVIS_BRANCH; else echo $TRAVIS_PULL_REQUEST_BRANCH; fi)
```

Instead of having users script this, I suggest adding an environment variable `$TRAVIS_CURRENT_BRANCH`.

**Comment by BanzaiMan** [comment]
What does "current" mean?

**Comment by breerly** [comment]
I document the issue here: https://graysonkoonce.com/getting-the-current-branch-name-during-a-pull-request-in-travis-ci/

**Comment by BanzaiMan** [comment]
Ah, I see. You want one environment variable that has different values depending on a push build or a PR build.

May I ask for the use case here?

**Comment by breerly** [comment]
We're using this value to push containers to Docker Hub after success:

https://github.com/yarpc/yarpc-go/blob/dev/.travis.yml#L89

Currently we have the above workaround in place in about 10 repos.

**Comment by BanzaiMan** [comment]
You can get to the value you want without `if` statement:

    ${TRAVIS_PULL_REQUEST_BRANCH:-TRAVIS_BRANCH}

This is `${TRAVIS_PULL_REQUEST_BRANCH}` if `${TRAVIS_PULL_REQUEST_BRANCH}` is nonempty (when the current job is a PR build) and `${TRAVIS_BRANCH}` if `${TRAVIS_PULL_REQUEST_BRANCH}` is empty.

**Comment by breerly** [comment]
That's still not great - I need the value multiple times, so it's either store that in a variable or duplicate it. This is something that should just be available without bash wizardry on the users begalf.

**Comment by breerly** [comment]
That's still not great - I need the value multiple times, so it's either store that in a variable or duplicate it. This is something that should just be available without bash wizardry on the users behalf.

**Comment by tomaszguzialek** [comment]
For all those wondering why this line doesn't work for them (as it did not work for me). One dollar sign is missing. The correct version:
```
${TRAVIS_PULL_REQUEST_BRANCH:-$TRAVIS_BRANCH}
```

Thanks for that trick, @BanzaiMan

**Comment by mirague** [comment]
It seems this was solved the way OP proposed in OndraM/ci-detector#17, what's your take on this @travis-ci?

**Comment by derek-j** [comment]
In building our Travis CI/CD system, I pushed from a feature branch not master.  This also fails when PUSHing can that be fixed.  A Single variable specifying the branch being built should be used ie  TRAVIS_BRANCH i don't see the purpose of ever giving it the tag name, when that can easily be retrieved.

**Comment by mikerott** [comment]
My solution:

```
    # We have to do some shinanigans to figure out what the source
    # branch is for this pull request.  Travis is not kindly giving
    # us an env var when it's a push to an already existing PR.
    echo "Running pull request build on Travis"
    repo_url=$(git config --get remote.origin.url)
    latestPRCommit=(${TRAVIS_COMMIT_RANGE//\.\.\./ })
    latestPRCommit=${latestPRCommit[1]}
    git clone ${repo_url} ${latestPRCommit}
    cd ${latestPRCommit}
    SOURCE_BRANCH=($(git branch -r --contains ${latestPRCommit}))
    export SOURCE_BRANCH=${SOURCE_BRANCH/origin\//}
    cd -
```

**Comment by sanket4373** [comment]
@breerly: I followed your way of getting the name of the current PR branch still facing problems though. posting my question here.

I want to get the name of the created PR in the travis logs. I want to use this PR name to build containers for my service every single time the PR is **created** or **updated** . In this case I am updating my PR.

 Posting the snippet of the travis build logs: 

```
$ export BRANCH=$(if [ ${TRAVIS_PULL_REQUEST} = "false" ]; then echo $TRAVIS_BRANCH; else echo $TRAVIS_PULL_REQUEST_BRANCH; fi)

The command "export BRANCH=$(if [ ${TRAVIS_PULL_REQUEST} = "false" ]; then echo $TRAVIS_BRANCH; else echo $TRAVIS_PULL_REQUEST_BRANCH; fi)" exited with 0.
0.01s$ echo "TRAVIS_BRANCH=$TRAVIS_BRANCH", "BRANCH=$BRANCH", "TRAVIS_PULL_REQUEST_BRANCH=$TRAVIS_PULL_REQUEST_BRANCH";
TRAVIS_BRANCH=develop, BRANCH=, TRAVIS_PULL_REQUEST_BRANCH=

```

As you can see`TRAVIS_PULL_REQUEST_BRANCH` shows empty output and `TRAVIS_BRANCH`shows the name of my base branch. 

Basically what is happening is this:

```
   # You open a PR with base `master`, and PR branch `foo`
    # During a PR build:
    #     TRAVIS_BRANCH=master
    #     TRAVIS_PULL_REQUEST_BRANCH=foo
    # During a push build:
    #     TRAVIS_BRANCH=foo
    #     TRAVIS_PULL_REQUEST_BRANCH=
```
Can someone kindly help me point out If I am missing something in the way I am performing this task?

**Comment by sanket4373** [comment]
Had to do this workaround to get the HEAD of the branch since even `git branch` was not helping to give the name of the origin branch.

```
git fetch origin refs/pull/"$TRAVIS_PULL_REQUEST"/head:"$TRAVIS_PULL_REQUEST"
git checkout "$TRAVIS_PULL_REQUEST"
```

This way got the HEAD of the branch and used it to build my container.

**Comment by StoneCypher** [comment]
@BanzaiMan - okay, so, this gets us the branch name

What about the repo name?  If I make a PR from my fork, the branch doesn't exist on master.  It needs to ask for `accountname:branchname`, rather than just `branchname`

**Comment by StoneCypher** [comment]
@BanzaiMan - I don't know if I feel safe at Travis if you've decided to throw away your backlog without fixing it

**Comment by mraible** [comment]
I'm using this in my `.travis.yml`.

```
  - export BRANCH=$(if [ "$TRAVIS_PULL_REQUEST" == "false" ]; then echo $TRAVIS_BRANCH; else echo $TRAVIS_PULL_REQUEST_BRANCH; fi)
  - echo "TRAVIS_BRANCH=$TRAVIS_BRANCH, PR=$TRAVIS_PULL_REQUEST, BRANCH=$BRANCH"
```

It worked great until I did a release.

```
export IONIC4J_REPO=https://github.com/oktadeveloper/generator-jhipster-ionic.git
export IONIC4J_BRANCH=$BRANCH
...
git clone $IONIC4J_REPO generator-jhipster-ionic
Cloning into 'generator-jhipster-ionic'...
remote: Counting objects: 859, done.
remote: Compressing objects: 100% (49/49), done.
remote: Total 859 (delta 16), reused 51 (delta 12), pack-reused 793
Receiving objects: 100% (859/859), 358.18 KiB | 3.94 MiB/s, done.
Resolving deltas: 100% (338/338), done.
cd generator-jhipster-ionic
if [ "$IONIC4J_BRANCH" == "latest" ]; then
    LATEST=$(git describe --abbrev=0)
    git checkout -b $LATEST $LATEST
elif [ "$IONIC4J_BRANCH" != "master" ]; then
    git checkout -b $IONIC4J_BRANCH origin/$IONIC4J_BRANCH
fi
fatal: 'origin/v3.2.0' is not a commit and a branch 'v3.2.0' cannot be created from it
```

Sorry for commenting on a closed ticket, but how do I adjust the script so this failure doesn't happen?

**Comment by StoneCypher** [comment]
@BanzaiMan - quite a few other valid tickets were also closed by this bot

**Comment by BenCoffeed** [comment]
I'm using an `after_success` command to call my AWX cluster to build code and ship the artifact to s3. I've used @mikerott 's workaround solution. However, it would be REALLY handy for scenarios such as this to have an origin slug / fork name and an origin branch name for PR's. Especially for changes pushed to existing PR's.

**Comment by augsteyer** [comment]
> Sorry for commenting on a closed ticket, but how do I adjust the script so this failure doesn't happen?

@mraible I think the solutions mentioned are forgetting that there is also another name a `$TRAVIS_BRANCH` can have.
From the docs:
> TRAVIS_BRANCH:
> for push builds, or builds not triggered by a pull request, this is the name of the branch.
> for builds triggered by a pull request this is the name of the branch targeted by the pull request.
> for builds triggered by a tag, this is the same as the name of the tag (TRAVIS_TAG).

So you are hitting the $TRAVIS_TAG issue where you tagging a repo produces a build. The solution above will also need to include this in the calculation, logically it will looks like this (psudo code):

```
if ($TRAVIS_EVENT_TYPE === 'pull_request') {
 $branch = $TRAVIS_PULL_REQUEST_BRANCH;
} elseif ($TRAVIS_TAG) {
  $tag = $TRAVIS_BRANCH; //handle tag commits
} else {
 $branch =  $TRAVIS_BRANCH;
}
```
Then you can just checkout the tag `git checkout tags/$tag -b $tag`

**Comment by CruiseMan** [comment]
@BanzaiMan @mikerott I know we all busy people, but maybe it's time to do something about 160 likes requested feature ?.

**Comment by fmohr** [comment]
@BanzaiMan joining in ... it would be a small step for you, but a giant leap for mankind.

**Comment by magthe** [comment]
Wouldn't it be lovely to have the PR branch name available in the webhook payload as well?

**Comment by crertel** [comment]
It looks like CircleCI already does this:

https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables

**Comment by leifmadsen** [comment]
It's 2019 and this would still be useful.

**Comment by achimnol** [comment]
Maybe `TRAVIS_ORIGIN_BRANCH` would be a good option.
I just encountered this case when I make PR against a feature branch that shares the same feature branch in a dependency package.
https://github.com/lablup/backend.ai-agent/blob/8ed5e04/.travis.yml#L28-L33

**Comment by gund** [comment]
This is ridiculous! I was using this $TRAVIS_BRANCH variable and trying to fix the issue thinking something is wrong in my code but it turns out that the variable name just does not represent what it says...

Yes I know that docs explain what this variable is doing but the name of it was so clear that I thought "what could possibly go wrong?...". Yeah naming IS the hardest problem in CS.

I'm also voting +1 for a variable that simply represents the "current" branch in all jobs.

**Comment by tillias** [comment]
I'm using heroku as a deployment destination and need this variable badly to avoid false deployment from pr/branches into staging environment. Voting up.

**Comment by StoneCypher** [comment]
this has been broken for four years.  it's not getting fixed.

travis has been in a holding pattern for years.  nothing is getting fixed.
